overlay_new.py

```diff
@@ -0,0 +1,197 @@
+from abc import ABC, abstractmethod
+from time import time
+
+from canvas import Canvas, Colour
+from data import Data
+from overlay import Overlay
+from textwrap import wrap
+from typing import Callable, Tuple
+
+class Drawable(ABC):
+    @abstractmethod
+    def draw_base(self, canvas: Canvas):
+        pass
+    @abstractmethod
+    def draw_data(self, canvas: Canvas, data: Data):
+        pass
+
+class DataField(Drawable):
+    """ A simple data field containing a large value below a title in a smaller
+        font. Text is right-aligned. """
+
+    title_size = 0.8
+    data_size = 1.5
+
+    width, data_height = Canvas.get_text_dimensions("100.0", data_size)
+    _, title_height = Canvas.get_text_dimensions("KPH", title_size)
+
+    spacing = 10
+
+    height = title_height + data_height + spacing
+
+    def __init__(self, title: str, value_func: Callable[[Data], str], coordinate: Tuple[int, int]):
+        """ `coordinate` specifies the bottom-left coordinate of the data field. """
+        self.title = title
+        self.value_func = value_func
+        self.title_coord = (coordinate[0] + DataField.width, coordinate[1] - DataField.spacing - DataField.data_height)
+        self.data_coord = (coordinate[0] + DataField.width, coordinate[1])
+
+    def draw_base(self, canvas: Canvas):
+        canvas.draw_text(self.title, self.title_coord, DataField.title_size, Colour.white, "right")
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        canvas.draw_text(self.value_func(data), self.data_coord, DataField.data_size, Colour.white, "right")
+
+class SpeedField(DataField):
+    """ A specialised version of DataField which prefers to display GPS speed,
+        but falls back to reed velocity if GPS is unavailable """
+
+    def __init__(self, coordinate: Tuple[int, int]):
+        value_func = lambda data: "{:.1f}".format(data["gps_speed"] or data["reed_velocity"])
+        super().__init__("", value_func, coordinate)
+
+    def draw_base(self, canvas: Canvas):
+        pass
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        self.title = "GPS KPH" if data["gps_speed"] else "REED KPH"
+        super().draw_base(canvas)
+        super().draw_data(canvas, data)
+
+class CentrePower(Drawable):
+    """ Displays the current and recommended power at the bottom-centre of the
+        screen.
+
+        If the current power is within 5% of recommended, it is coloured green,
+        otherwise red. If no recommended power is available, it is coloured
+        black. """
+
+    power_size = 2.5
+    rec_power_size = 1.5
+    _, rec_power_height = Canvas.get_text_dimensions("500w", rec_power_size)
+    spacing = 10
+
+    power_tolerance = 0.05
+
+    def __init__(self, screen_w: int, screen_h: int):
+        self.power_coord = (screen_w // 2, screen_h - 30 - CentrePower.spacing - CentrePower.rec_power_height)
+        self.rec_power_coord = (screen_w // 2, screen_h - 30)
+
+    def draw_base(self, canvas: Canvas):
```

Comment:
I kind of feel like it doesn't violate ISP. In the article you cited, it uses examples where calling a some method on a class which doesn't "need" the method but implements it because it has to, is invalid. Borrowing their example, calling `orderFries()` on a `BurgerOrderService` object.

However, I argue that this does not violate ISP because calling `drawBase` on CentrePower is not invalid - there is simply nothing to be done in this instance. But say we wanted to put a permanent white rectangle behind the data - we could just add it to `drawBase`. It is equally valid to want to draw something to the base layer or to not require anything to be drawn. However, there is no logical case where it would make sense for a service soley dedicated to ordering burgers to order fries - which is why that violates ISP, but my case doesn't. That's what I'm thinking anyway.

Alternatively, I have two options. Split `Drawable` into two separate interfaces, `BaseDrawable` and `DataDrawable`. Or, I could simply make `drawBase` not abstract, and leave it with a default empty implementation. I'd lean towards the latter, as that way we won't have to treat things that may be drawn on the base layer differently to those that can't when coding the overlay.

What do you think?

Reply:
I see what you mean by the ISP, with the fact that we don't need anything drawn in the base layer for it. If we were to implement a way to manually render the base canvas, then this wouldn't be an issue as we would be able to utilise the `draw_base` method here.

Not a big fan of splitting the two interfaces in this instance as it would require us to know what should and shouldn't be on the base canvas.

Reply:
I know what you're saying, but I also feel like manually redrawing the base layer defeats its purpose, with the base layer being for static content that is only rendered once, and the data layer for dynamic content.

> Not a big fan of splitting the two interfaces in this instance as it would require us to know what should and shouldn't be on the base canvas.

Removing `draw_base` from the interface would also require us to manually keep track of what should be drawn to the base canvas. Removing `draw_base` from the interface and splitting the interface into two separate interfaces is kinda doing the same thing, as I see it, just in the latter case, everything we can call `draw_base` on is still logically grouped. Do you see it differently?

Reply:
Side note... I do wonder if we should even have separate base and data overlays. Sure, we save time by not rewriting a bunch of titles every frame with the base overlay, but it also means we have to merge two overlays with the video feed rather than just one... Would be super interesting to investigate how much of a performance increase we really get from this increased complexity

Reply:
> Would be super interesting to investigate how much of a performance increase we really get from this increased complexity

IIRC it helped MASSIVELY. This was not with open cv though, so maybe it will have changed?

Reply:
> I know what you're saying, but I also feel like manually redrawing the base layer defeats its purpose, with the base layer being for static content that is only rendered once, and the data layer for dynamic content.

I don't think that is a problem where we could instead say that the base layer is for content that isn't changed _often_ and the data layer to be data that is expected to change upon every data call.

How about this suggestion:
A `is_title_static` argument that takes in a `boolean` which is passed into `DataField`. 

`DataField`
```python
def draw_data(self, canvas: Canvas, data: Data):
	if not is_title_static:
		self.draw_base(canvas)
    canvas.draw_text(self.value_func(data), self.data_coord, DataField.data_size, Colour.white, "right")
```

`SpeedField`
```python
    def draw_data(self, canvas: Canvas, data: Data):
        self.title = "GPS KPH" if data["gps_speed"] else "REED KPH"
        super().draw_data(canvas, data)
````

The main reason for this would be that we don't need to pass in the logic of rendering the title to whoever inherits the `DataField` class. It just looks weird if the user of the class calls `super.draw_base()` from its `draw_data` method.

Reply:
Check my most recent commit, is this what you meant? Also interested if you know a better way of doing the `DataField.draw_base(self, canvas)` thing.

Reply:
Yeah that's what I wanted! ATM not too sure how else I would do it tbh. But for now, the `draw_base` and `draw_data` are seperated in the child components